File: ios/Cove/Flows/SettingsFlow/SettingsScreen/WalletSettingsContainer.swift
//
//  WalletSettingsContainer.swift
//  Cove
//
//  Created by Praveen Perera on 12/5/24.
//

import Foundation
import SwiftUI

struct WalletSettingsContainer: View {
    @Environment(AppManager.self) var app

    // args
    let id: WalletId
    let route: WalletSettingsRoute

    // private
    @State private var manager: WalletManager? = nil
    @State private var error: String? = nil

    func initOnAppear() {
        do {
            let manager = try app.getWalletManager(id: id)
            self.manager = manager
        } catch {
            self.error = "Failed to get wallet \(error.localizedDescription)"
            Log.error(self.error!)
        }
    }

    @ViewBuilder
    func WalletSettingsRoute(manager: WalletManager, route: WalletSettingsRoute) -> some View {
        switch route {
        case .main:
            WalletSettingsView(manager: manager)
        case .changeName:
            EmptyView()
        }
    }

    @ViewBuilder
    var LoadingOrError: some View {
        Group {
            if let error {
                Text(error)
            } else {
                FullPageLoadingView()
            }
        }
        .task {
            guard let error else { return }
            Log.error(error)
            try? await Task.sleep(for: .seconds(5))
            app.resetRoute(to: .listWallets)
        }
        .onAppear(perform: initOnAppear)
    }

    var body: some View {
        if let manager {
            WalletSettingsRoute(manager: manager, route: route)
        } else {
            LoadingOrError
        }
    }
}

File: android/app/src/main/java/org/bitcoinppl/cove/WalletSettingsContainer.kt
package org.bitcoinppl.cove

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import org.bitcoinppl.cove_core.*
import org.bitcoinppl.cove_core.types.*

/**
 * wallet settings container - lazy loads WalletManager for wallet settings
 * ported from iOS WalletSettingsContainer.swift
 */
@Composable
fun WalletSettingsContainer(
    app: AppManager,
    id: WalletId,
    route: WalletSettingsRoute,
    modifier: Modifier = Modifier,
) {
    var manager by remember(id) { mutableStateOf<WalletManager?>(null) }
    val tag = "WalletSettingsContainer"

    // lazy load wallet manager
    LaunchedEffect(id) {
        try {
            android.util.Log.d(tag, "getting wallet $id")
            manager = app.getWalletManager(id)
        } catch (e: Exception) {
            android.util.Log.e(tag, "failed to load wallet", e)
            app.alertState =
                TaggedItem(
                    AppAlertState.General(
                        title = "Error!",
                        message = "Unable to load wallet: ${e.message}",
                    ),
                )
        }
    }

    // render
    when (val wm = manager) {
        null -> {
            Box(
                modifier = modifier.fillMaxSize(),
                contentAlignment = Alignment.Center,
            ) {
                CircularProgressIndicator()
            }
        }
        else -> {
            when (route) {
                WalletSettingsRoute.MAIN -> {
                    // TODO: implement WalletSettingsScreen with manager
                    Box(
                        modifier = modifier.fillMaxSize(),
                        contentAlignment = Alignment.Center,
                    ) {
                        androidx.compose.material3.Text("Wallet Settings - TODO")
                    }
                }
                WalletSettingsRoute.CHANGE_NAME -> {
                    // TODO: implement change wallet name screen
                    Box(
                        modifier = modifier.fillMaxSize(),
                        contentAlignment = Alignment.Center,
                    ) {
                        androidx.compose.material3.Text("Change Name - TODO")
                    }
                }
            }
        }
    }
}
